✨공통 컴포넌트 구현

Legitgoons opened the pull request:
## 작업 이유
- 작업 속도 및 유지보수성 향상을 위해 공통 컴포넌트 구현

<br/>

## 작업 사항
### 1️⃣ 컴포넌트 구현
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/7a3ab043-d66c-4a66-8b5b-8135d0ae65e4)

#### BasicButton / ActiveButton
- 복잡성 제어를 고려해 isDisabled 속성에 따라 별개의 컴포넌트로 구현
- interface는 type.ts에서 상속받아 사용
- 선언된 class들을 prop로 입력하도록 설정
```tsx
//BasicButton.tsx
export type BasicStyle = 'defalut' | 'confirm-cancle' | 'bsm-cancle' | 'bsm-option';
interface BasicButtonProps extends ButtonProps {
  styleClass: BasicStyle;
}
<button onClick={onClick} className={styleClass}>
```
#### Modal
- index.html에 overlays div태그를 생성
```html
// index.html
  <body>
    <div id="root"></div>
    <div id="overlays"></div>
  </body>
```
- ModalOverlay내에 createPortal을 사용해 Modal이 상위요소의 위치에 영향을 받지 않도록
- BottomSheetModal의 경우를 고려해, styleClass prop에 따라 다른 classname이 설정되도록 함
```tsx
type ModalOverlayClass = 'bottom-sheet' | 'modal';

export const ModalOverlay = ({ styleClass, children, onClose, }: ModalOverlayProps) => {
  return (
    <>
      {createPortal(<Backdrop onClick={onClose} />, portalElement)}
      {createPortal(
        <figure className={styleClass}>{children}</figure>,
        portalElement,
//...
```
- 이를 다른 Modal들에서 사용
```tsx
// ConfirmModal.tsx
<ModalOverlay styleClass='modal' onClose={onClose}>
  <div className='confirm-modal'>
// ... 
```
#### BottomSheetModal
```tsx
{options.map((option, index) => (
  <Fragment key={index}>
    <BasicButton onClick={option.onClick} styleClass='bsm-option'>
      {option.label}
    </BasicButton>
    {index < options.length - 1 && <hr className='option-divider' />}
    // 마지막 option을 제외한 option에만 hr태그 생성
  </Fragment>
))}
```
```scss
.optionDivider {
  height: 1px; // 0.5px 사용시 이슈 있음
}
```
#### PageHeader
- {page} 중앙 정렬을 위해 right padding 44px 사용

### 2️⃣ vite-plugin-svgr 사용
- svg를 컴포넌트로 사용하기 위해 vite-plugin-svgr 설치 및 설정
- 아래와 같이 사용 시 svg를 컴포넌트로 사용 가능
```ts
import {svgComponent} from '{경로}/{svgName}.svg?react';
```

### 3️⃣ 기타
- reset.scss를 사용해 style sheet 초기화
- icon을 담기 위해 shared/icon 폴더 생성
  - logo를 담당하던 기존의 icon.svg는 혼동의 여지가 있기에  logo.svg로 이름 변경
- 네이밍 룰에 따라 폴더 이름 수정
  - shared/ui/FNB -> fnb로 변경
- fonts.scss 파일 font들에 사이즈를 주석으로 작성

<br/>

## 리뷰어가 중점적으로 확인해야 하는 부분
- [x] 컴포넌트가 적절하게 구현되었는지
- [x] 컴포넌트의 이름이 적절한지
- [x] vite-plugin-svgr를 사용해 svg를 컴포넌트로 사용하는 것을 인지하였는지
- [x] 아래의 이슈를 인지하였는지
<br/>

## 발견한 이슈
- Design팀에서 BottomModal의 구분선을 0.5px로 설정, 하지만 chrome에서 0.5px가 정상적으로 출력 되지 않음
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/fdce4cd9-8096-4884-be84-bf3891c90048)

BangDori committed:
⚙️ CI 파이프라인 및 husky 스크립트 수정 (#13)

- husky install 수정
- CI 파이프라인 스크립트 및 이미지 수정

Legitgoons committed:
⚙️반응형 사이즈 설정 (#14)

* feat: 반응형 레이아웃을 위한 mixin responsive-dimensions 구현

* feat: 공용 style의 쉬운 사용을 위해 _index.scss 생성

Legitgoons committed:
fix: FNB 폴더 네이밍 룰에 따라 fnb로 수정

Legitgoons committed:
feat: fonts의 사이즈를 주석으로 작성

Legitgoons committed:
feat: BaseButtonProps와 이를 상속한 ActiveButtonProps 작성

Legitgoons committed:
feat: ConfirmButton, ConfirmCancleButton 구현

Legitgoons committed:
feat: FollowButtonLg, FollowButtonSm 구현

Legitgoons committed:
feat: Profile 구현

Legitgoons committed:
feat: shared/ui/button/buttonType -> types로 이름 변경

Legitgoons committed:
feat: svg를 컴포넌트로 사용하기 위해 vite-plugin-svgr 설치

Legitgoons committed:
feat: '*.svg' 타입 선언을 위해 svg.d.ts파일 생성

Legitgoons committed:
feat: vite-plugin-svgr를 tsconfig.json과 vite.config.ts에 추가

Legitgoons committed:
feat: back-icon.svg 파일 추가

Legitgoons committed:
feat: 기존의 icon.svg은 logo.svg로 이름 변경

Legitgoons committed:
feat: Normalize로 기본 stylesheet reset

Legitgoons committed:
fix: fnb 폴더명 변경 적용

Legitgoons committed:
fix: confirmButtonActive 매개변수 수정

Legitgoons committed:
feat: Backdrop 컴포넌트 구현

modal은 z-index 10, backdrop은 z-index 5로 구현

Legitgoons committed:
feat: ConfirmModal 구현

Legitgoons committed:
feat: BottomModal 구현

Legitgoons committed:
feat: modal/types.ts 구현

Legitgoons committed:
feat: PageHeader 컴포넌트 구현

{page} 중앙 정렬을 위해 우측에 dummy div 태그 사용

Legitgoons committed:
feat: Profile.scss 내 class 정렬

Legitgoons committed:
feat: BottomSheetModal 이름 변경 및 bottom-sheet 폴더로 이동

Legitgoons committed:
feat: ConfirmModal description -> content로 수정

Legitgoons committed:
feat: ConfirmModal mixin 생성, confirmReportModal 분기처리하도록 구현

Legitgoons committed:
feat: 변경 가능성 고려 ConfirmReportModal 구현

Legitgoons committed:
feat: PageHeader 사용하지 않는 태그 삭제

Legitgoons committed:
fix: 이름 변경 후 남아있는 FNB 폴더 삭제

psychology50 commented:
`Readme` 수정과 같은 작업은 별도의 PR로 분리했어도 좋을 것 같아요.  
`pipeline`을 수정하는 작업도 마찬가지구요.  

관심사가 분리되지 않은 커다란 하나의 PR보다, 적절히 관심사가 분리된 많은 PR이 리뷰어의 집중을 흐트리지 않아서 오류를 초기에 잡아낼 수 있게 되더라구요.  
물론 너무 사소한 작업이라 사전에 합의를 보고 진행해도 되긴 하겠지만 ㅎㅎ 저도 가끔 그리 합니다.  

집중적으로 확인해야 할 항목은 병준님께 토스~

BangDori commented on svg.d.ts:
P4: index.d.ts로 작성해두고 추후에 다른 앰비언트 타입을 추가해주는 것도 괜찮지 않을까요?

많진 않겠지만 svg 뿐만 아니라 추후에 다른 앰비언트 타입을 작성할 경우가 발생할 수 있을 것 같은데 이에 대한 앰비언트 타입 선언을 위해 각 파일을 생성해서 관리하는 건 불 필요할 수 있다고 생각이 들어요.

그리고 앰비언트 파일이 생길 때 마다, tsconfig.json에 명시된 `include` 속성에 추가해줘야 하는 번거로움도 생길 수 있을 것 같구요.

Legitgoons replied:
맞네용 index.d.ts로 수정하겠슴다!

BangDori commented on src/shared/ui/bottom-sheet/BottomSheetModal.scss:
```diff
@@ -0,0 +1,45 @@
+@use '../../styles/_index.scss' as *;
```
P4: alias 경로 적용해서 작성해주는게 더 좋은 것 같아요!

스타일 경로가 복잡해서라기보다는 alias 경로 적용해서 하는게 해당 `_index.scss` 위치 파악하기도 좋을 것 같아서, alias 경로 적용하는 게 더 좋을 것 같아요!

Legitgoons replied:
그렇네요! 수정하겠습니당

Legitgoons replied:
넵 수정하겠습니당

BangDori commented on src/shared/ui/bottom-sheet/BottomSheetModal.scss:
```diff
@@ -0,0 +1,45 @@
+    @include responsive-dimensions(280px, 40px);
+    background-color: $gray1;
+    color: $gray5;
+    border: none;
```
P3: button 태그 자주 사용하게 될 것 같은데, button 태그의 고유 border 속성을 매번 none 처리해주는 건 다소 번거롭게 느껴질 수도 있을 것 같아서 전역 설정으로 button 태그에 대해 `border: none` 처리해주는게 좋을 것 같아요!

Legitgoons replied:
넵 이 부분 shared/styles/global.scss 파일 하나 더 만들어서 거기서 처리하겠습니다!

BangDori commented on src/shared/ui/profile/Profile.tsx:
```diff
@@ -0,0 +1,19 @@
+import './Profile.scss';
+
+interface ProfileProps {
+  profileImage: string;
```
P3: 모든 사용자가 자신의 프로필 이미지를 등록해두진 않았을 거라고 생각해요. 사용자의 프로필 이미지가 존재하지 않는 경우는 어떻게 될까요?

1. 서버 측에서 사용자 기본 이미지 URL을 받는다.
2. 프론트측에서 기본 이미지를 저장해두고, 프로필 이미지가 없는 경우 기본 이미지를 표시해준다.

우선 서버와 이에 대해 이야기를 나눠봐야 할 것 같네요!

Legitgoons replied:
이미지가 없는 사람의 숫자가 상당할테니, 저는 2번이 좀 더 적합하다고 생각합니다!

BangDori replied:
음 네넵! 방금 재서님이랑 얘기해봤는데, 프로필 이미지가 없는 경우에는 `null` 값으로 저장이 된다고 하네요.

그럼 프로필 이미지의 타입을 받아올 때, `string | null`로 저장해두고 타입이 `null` 이라면 프로젝트 내에 존재하는 기본 이미지를 렌더링해주는 방향으로 하시면 될 것 같습니다.

psychology50 replied:
저는 기본적으로 `null`을 응답값에 포함시키는 게 굉장히 불안 요소라고 생각합니다.  
이는 API 뿐만 아니라 기본적인 프로그래밍 원칙에 위배한다고 보구요.  

그래서 성공 응답 데이터에 아무것도 없는 경우에도 `"data": null`이 아닌, `"data": {}`를 반환합니다.  

프로필 이미지처럼 문자열 데이터가 데이터 베이스에 `null`로 저장되어 있더라도, 응답에는 `""`로 반환합니다.  
숫자의 경우엔 음수값을 보낸다던가 하는 식으로 처리해요.  
그마저도 안 된다면 플래그 값을 사용해서 응답 데이터를 반환할 예정입니다.

BangDori commented on src/shared/ui/profile/Profile.scss:
```diff
@@ -0,0 +1,32 @@
+
+    .userName {
+      color: $gray7;
+      margin: 0;
```
P3: `h` 태그와 `p` 태그에 대해서도 매번 `margin: 0`을 적용하기 번거로울 수 있으니, 이에 대해서도 전역으로 처리해두는게 편리할 것 같아요!

Legitgoons replied:
넵 마찬가지로 shared/styles/global.scss에서 처리하겠습니다!

BangDori commented on src/shared/ui/modal/ConfirmReportModal.tsx:
P2: 신고 모달창의 경우에는 아마 아래 이미지처럼 고정될 것 같아요!

![image](https://github.com/CollaBu/pennyway-client-webview/assets/44726494/3ebf492b-a588-4154-8f96-d9296b39180c)

`children`을 받는게 아니라, 위 이미지에 맞춰서 `ConfirmReportModal` 컴포넌트를 만들면 될 것 같아요!!

Legitgoons replied:
넵 이 부분 추가 구현하도록 하겠습니당

Legitgoons replied:
병준님 생각해보니 이게 재사용될 부분이 없이 신고하기에만 사용되는 컴포넌트라면, 공통 컴포넌트 branch가 아니라 추후 신고 기능을 만들 때 함께 구현하는게 나을 것 같습니다.

BangDori commented on src/shared/ui/modal/ConfirmReportModal.tsx:
```diff
@@ -0,0 +1,33 @@
+  onCloseMsg,
+}: ConfirmReportModalProps) {
+  return (
+    <div className='confirmReportModal'>
```
P2: 신고 모달창의 경우 확인 버튼을 클릭했을 때, 서버로 데이터를 전달하는 Action이 발생하게 되서 form 태그로 만드는 게 더 좋지 않을까요??

Legitgoons replied:
form 태그를 잊고있었네요. 수정하겠습니다!

BangDori commented on src/shared/ui/bottom-sheet/BottomSheetModal.tsx:
```diff
@@ -0,0 +1,29 @@
+  options: Option[];
+}
+
+export default function BottomSheetModal({ options }: BottomSheetModalProps) {
```
P4: BottomSheetModal에 옵션을 `props`로 받는 이유가 있으실까요?

Legitgoons replied:
여러 군데에서 재사용 될 수 있기에 하나의 option을 고정적으로 가지는 것 보단 props로 options을 전달받는 것이 적절하다고 생각했습니다.

BangDori replied:
컴포넌트의 재사용 여부와 option이 변경될 여지가 있는지에 대한 의문이 생기는 것 같습니다.

저희 프로젝트의 피그마를 살펴보면 피드에서 사용되는 BottomSheetModal이 사용되는 부분은 해당 피드에 대한 공유하기 버튼을 클릭했을 때 밖에 없는 것 같아요.

![image](https://github.com/CollaBu/pennyway-client-webview/assets/44726494/c1c5065c-cee4-46fd-bd56-15c636650c6d)

혹시 제가 찾지 못한 BottomSheetModal이 있을까요?

Legitgoons replied:
아 위의 컴포넌트는 피드 공유 부분에서만 사용되기에 공통 컴포넌트로 보기 애매하다고 판단해 따로 구현하지 않았습니다. 

구현된 BottomSheetModal은 피드 작성시 사진 첨부와 댓글 수정/삭제 등에 사용되는 컴포넌트입니다!
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/8295af07-1bb5-43f3-b556-dbc9b208e6fd)
![image](https://github.com/CollaBu/pennyway-client-webview/assets/101088491/556e553f-00dd-4014-a958-e0255ee230ee)

BangDori replied:
아! 이 부분에 대한 컴포넌트였꾼요.

제가 잘못 이해하고 있었네요, 확인했습니다!

BangDori commented on public/assets/icon/back-icon.svg:
P3: svg를 src 가 아닌, public에서 관리하는 이유가 궁금합니다!

Legitgoons replied:
svg와 font같은 파일들은 public에 저장해야한다고 알고 있었는데 찾아보니 그게 아니였네요;
src/shared/icon 폴더 생성하고 이동시키겠습니다!

BangDori replied:
엇 public에 저장하면 안되는 이유에 대해 공유 가능할까요?

Legitgoons replied:
주된 이유는 컴파일 단계에서의 에러 파악 가능 여부입니다.
이 블로그를 참조했습니다! https://jaypedia.tistory.com/298

BangDori replied:
아하 그렇군요! 좋은 정보 감사드립니다 ㅎㅎ

BangDori commented on src/shared/styles/fonts.scss:
P5: 이거 저는 개인적으로 한 줄로 작성하고, 뒤에 px 주석으로 추가해주는게 더 가독성 좋은 것 같아요!

`.h1bold { @include font-style(bold, 1.5, 130%); } // 24px`

Legitgoons replied:
1. 저도 그렇게 생각하지만, prettier로 인해 강제 변환되었습니다.
2. 원인으로 추측되는 printWidth 속성을 수정해봤는데, 숫자를 몇으로 늘려도 현재 형태로 되돌아가네요
3. 일단 이 부분은 추후에 수정해보겠습니다.

BangDori replied:
음 아니면 bold, semibold, medium 등에 대한 폰트 사이즈를 나타내는 거니까 상단 부분에 따로 주석 처리를 통해 나타내는 것도 괜찮을 것 같아요.

```
/**
 * bold: 18px
 * semibold: 14px
 * ...
 */

@mixin font-style($weight, $size, $line-height) {

// ....
```

Legitgoons replied:
mixin 위에 작성하는 방향으로 가면 오히려 더 가독성이 떨어질 것 같아서 이 방안은 좀 애매해보입니다. 차라리 font class 위쪽에 작성하는 건 어떨까요?
```scss
// 24px
.h1bold {
  @include font-style(bold, 1.5, 130%);
}
//....
```

BangDori replied:
아, 죄송합니다 bold, semibold, medium에 대한 폰트라고 잘못 이해해버렸네요.

의찬님 말씀대로 확실히 위로 가면, 모든 클래스에 대한 폰트 사이즈를 정의해줘야 하고 오히려 가독성이 떨어질 수 있을 것 같네요. 그냥 .prettierignore 를 만들어서, fonts에 대한 prettier를 적용하지 않도록 하는게 좋을 것 같네요.

그리고 이건 개인적인 생각이지만 하나의 폰트 클래스를 이해하기 위해 두 라인을 읽는 것 보다는 한 라인만 읽고 이해하도록 하는 게 더 좋은 것 같아서, 클래스 뒤에 주석으로 폰트 사이즈를 작성해주는 게 더 좋을 것 같아요.

Legitgoons replied:
넵 저도 한 라인만 읽도록 하는게 훨씬 가독성이 좋다고 생각합니다. prettierignore를 만들고 fonts.scss를 추가하도록 하겠습니다!

BangDori commented on src/shared/ui/button/FollowButtonLg.tsx:
```diff
@@ -0,0 +1,17 @@
+import './followButton.scss';
+import { ActiveButtonProps } from './types';
+
+interface FollowButtonLgProps extends ActiveButtonProps {}
```
P3: 확장 될 정보가 없다면, `ActiveButtonProps`를 `import` 해올 때 `as`를 활용해서 `FollowButtonLgProps`로 받아오는 게 더 좋지 않을까요?

Legitgoons replied:
이 부분은 미처 생각하지 못했네요. 수정하겠습니다!

BangDori commented on src/shared/ui/button/ConfirmButton.tsx:
```diff
@@ -0,0 +1,19 @@
+
+export default function ConfirmButton({
+  onClick,
+  isActive,
```
P4: `isActive` props가 꼭 필요한 것인가요?

Legitgoons replied:
UI에서 ConfirmButton의 동작을 막으면서 색상도 같이 변경되는 경우가 있습니다.
이 경우 자체적인 상태값을 가지거나 전역 상태로 관리하기보다는 props로 내려받는 것이 더 낫다고 판단했습니다.

BangDori replied:
버튼을 막는다는 것은 `disabled`가 true로 설정된다는 것을 의미하는 것 같습니다.

그렇다면 클래스를 생성하는 것이 아니라, 해당 클래스에 대해 `disabled`가 true일 경우 어떠한 스타일이 적용되도록 나타내는 방법은 어떠신가요?

Legitgoons replied:
disabled를 가지고 판별한다고 하더라도, 그 disabled를 판별하는 prop이 또 필요하지 않을까요...?
이 경우 isDisabled로 이름을 변경한다면 좀 더 직관적일 것 같긴 합니다.

BangDori replied:
그렇군요! 제가 제안한 방법은 제 생각일 뿐이라, 반영하셔두 되고 반영하지 않으셔도 상관은 없습니다! ㅎㅎ

Legitgoons replied:
넵 그럼 prop명을 isDisabled로 수정하고, button 태그에도 추가하겠습니다!

BangDori commented on src/shared/ui/button/FollowButtonSm.tsx:
P3: `shared/ui/Button` 내부에 버튼 컴포넌트가 컴포넌트명을 통해 어떤 역할을 하는지 잘 유추되긴 하지만, 거의 대부분의 버튼이 스타일만 다를 뿐 컴포넌트 내부가 동일한 형태를 띄고 있는 것 같아요.

그렇다면 굳이 각각의 컴포넌트로 분리하지 않고, 클래스명을 통해 해당 버튼을 유추할 수 있지 않을까요?

```html
<Button className='follow-btn-sm' />
```

Legitgoons replied:
지나친 확장성으로 인해서 만능 컴포넌트가 되는 것을 경계하려고 했지만, 버튼 컴포넌트끼리 중복된 부분이 많은것도 사실입니다.

활성화/비활성화 속성 여부에 따라 ActiveButton / BasicButton 두 가지로만 나누고, union type으로 props를 받아 스타일 클래스를 설정해주는 건 어떨까요?

BangDori replied:
ConfirmButton.tsx Review에서 진행한 내용과 동일한 내용인데, 만약 활성화/비활성화를 `disabled`로 나타내고 이를 props로 받아온다면 그냥 하나의 버튼만으로도 충분할 것 같네요!

Legitgoons replied:
그렇게 한다면 상태가 없는 버튼도 isDisable  props를 가지게 되기에, 헷갈릴 수 있다고 생각합니다. optinal로 준다면 반대로 상태가 있는 버튼이 헷갈릴 수 있구요.

또 props에 따라 처리할 사항도 늘어나 유지보수에도 악영향이 있을 수 있습니다. ActiveButton과 BasicButton으로 나누는게 나을 것이라 생각합니다.

BangDori replied:
저는 오히려 이 부분에 대해서는 버튼을 하나로 두는 게 더 낫다고 생각을 해요.

shared/ui 폴더에 들어오는 컴포넌트들의 경우 재사용의 목적이 크지, 어떠한 역할을 나타내는 컴포넌트라는 생각은 사실 잘 들지 않거든요. 그리고 활성화의 유무로 인해 컴포넌트를 분리하게 되면, 추후에 또 어떠한 차이로 인해 다른 스타일이 적용된다거나 그랬을 때 이에 대한 버튼을 또 다시 분리해야 하는 문제에 놓일 수 있을 것 같아요.

실제 FSD 아키텍처를 기반으로 만들어진 [nukeapp](https://github.com/noveogroup-amorgunov/nukeapp/blob/main/src/shared/ui/Button/Button.tsx)과 [sudoku-effector](https://github.com/Shiyan7/sudoku-effector/blob/master/src/shared/ui/button/ui.tsx) 프로젝트들은 버튼이라는 하나의 컴포넌트에서 모든 props를 받아서 처리하고 있어요.

Legitgoons replied:
```tsx
interface ButtonProps {
  onClick: () => void;
  styleClass: 'confirm' | 'confirm-cancle' | 'follow-sm' | 'follow-lg'; 
  isDisabled?: boolean;
  children: React.ReactNode;
}

export default function Button({
  onClick,
  styleClass,
  isDisabled = false,
  children,
}: ButtonProps) {
  const sytleClassname = isDisabled ? styleClass : `${styleClass}-disabled`;
  return (
    <button
      onClick={onClick}
      className={`${sytleClassname} h4semi`}
      disabled={isDisabled}
    >
      {children}
    </button>
  );
}
```
- 공통 Button 컴포넌트를 구현해보았습니다. 하지만 여전히 의문이 있습니다.
- 공통 컴포넌트의 목표는 적절한 재사용을 통한 개발시간 단축과 유지보수성 개선입니다.
  - 재사용만을 위해서 공통 컴포넌트를 만든다면 빈 DOM Element를 지향하게 되지 않을까요?
- 따라서 어느 정도 역할을 명확히 해 줄 필요가 있다고 느낍니다.
  - 공통 컴포넌트가 많은 역할을 수행한다면, 사용자 입장에서는 신경써야 할 props가 늘어나게 됩니다.
  - 추후에 명확한 역할 차이가 있는 버튼이 생긴다면, 저는 이것 역시도 분리하는게 낫다고 생각합니다.

BangDori replied:
그렇다면 우선 의찬님의 말씀처럼 활성화 여부에 따른 버튼을 나누어 두고, 추후에 역할 차이가 있는 버튼이 생기게 되었을 때 다시 이야기를 해보는 것도 좋을 것 같습니다.

BangDori reviewed (changes requested):
라인은 많지 않은데, 새롭게 추가된 컴포넌트랑 스타일시트가 많아서 생각보다 오래 걸렸네요!!

Legitgoons committed:
feat: 전역 style 선언을 위한 global.scss 생성 및 적용

Legitgoons committed:
feat: Profile styleclass props와 일치하게 이름 수정

Legitgoons committed:
feat: @use '/styles/_index.scss' 경로 파악을 위해 절대경로로 수정

Legitgoons committed:
feat: 전역 스타일 설정과 중복된 스타일 설정 제거

Legitgoons committed:
feat: 추후 다른 앰비언트 타입 작성을 고려해 svg.d.ts -> index.d.ts로 수정

Legitgoons committed:
feat: 코드의 가독성을 위해 fonts.scss에서 prettier 비활성화

Legitgoons committed:
feat: confirmReportModal form 태그로 수정

Legitgoons committed:
feat: 컴파일 단계에서의 에러 파악을 위해 svg icon src/shared/icon으로 이동

Legitgoons committed:
feat: isDisabled를 prop으로 받는 ActiveButton 컴포넌트 구현

Legitgoons committed:
feat: BasicButton 컴포넌트 구현

Legitgoons committed:
feat: ButtonProps 수정

Legitgoons committed:
feat: 기존의 Button 컴포넌트들 제거

Legitgoons committed:
feat: ConfirmModal 컴포넌트 내 Button들 수정

Legitgoons committed:
feat: BottomSheetModal 컴포넌트 내 버튼이 BasicButton 사용하도록 수정

Legitgoons committed:
feat: naming rule에 따라 style class들을 케밥케이스로 수정

BangDori commented on src/shared/ui/profile/Profile.tsx:
```diff
@@ -0,0 +1,19 @@
+export default function Profile({ profileImage, name, content }: ProfileProps) {
+  return (
+    <div className='profile'>
+      <img src={profileImage} className='profile-image' />
```
P3: 혹시 프로필 이미지 없는 경우에 기본 이미지로 표시되도록 하는 건 `props`로 받는 건가요? 아니면 처리가 아직 안된 건가요?

Legitgoons replied:
기본 이미지 표시는 PM/디자인 팀과 논의하고 기본 이미지가 결정된 후 처리하려고 했는데, 생각해보니 일단 로직만이라도 간단하게 구현해놓는게 나을 것 같네요. 추가하겠습니다!

BangDori commented on src/shared/ui/modal/ConfirmReportModal.tsx:
```diff
@@ -0,0 +1,32 @@
+
+import { BaseModalProps as ConfirmReportModalProps } from './types';
+
+export default function ConfirmReportModal({
```
P3: 이 부분 생각을 해보니, 의찬님 말씀대로 신고하기 기능 구현할 때 만드는게 맞는 것 같아요. 궁금한 점이 있는데, 저 같은 경우에는 모달창을 구현할 때 React Portal을 이용했었는데 의찬님은 모달창 어떻게 표시하셨나요?

Legitgoons replied:
저는 보통 단순하게 pixed로 컴포넌트를 위치시키고 modal 컴포넌트 / backdrop 컴포넌트를 따로 구현해서 동시에 실행시키는 방식으로 했었습니다. 그런데 지금 React Portal에 대해서 찾아봤는데 이게 더 좋은 것 같은데요...? Modal과 BottomSheet가 React Portal을 사용해서 동작하도록 수정하겠습니다 ㅋㅋㅋㅋ

BangDori commented on src/shared/ui/button/BasicButton.tsx:
P4: `styleClass`의 타입이 현재 `'confirm-cancle' | 'bsm-cancle' | 'bsm-option'` 타입으로 적용되어 있는데, 일반적인 버튼의 경우에는 해당 컴포넌트를 사용하지 못하는 것인가요?

Legitgoons replied:
일반적인 경우를 고려해서 default style class를 추가적으로 만들어두겠습니당

BangDori commented on src/shared/ui/button/BasicButton.scss:
P4: 바로 위 리뷰와 동일한 내용이긴 한데, 해당 버튼에 들어올 수 있는 클래스들을 여기서 정의하게 되버리면 해당 클래스들만 사용하거나 혹은 클래스가 추가될 때 마다 여기에 정의해야 할 텐데 그럼 오히려 복잡해질 수 있지 않을까요?

Legitgoons replied:
이걸 생각해보느라고 답변이 좀 늦었네요. 우선 개선방안에 대해서는 다음과 같이 따로 union type으로 분리해서 사용하는 것이 사용자가 해당 타입이 유니온으로 정의되어 있다는 것을 좀 더 직관적으로 아는데 도움이 될 것 같습니다.
```ts
export type BasicStyle = 'confirm-cancle' | 'bsm-cancle' | 'bsm-option';

interface BasicButtonProps extends ButtonProps {
  styleClass: BasicStyle;
}
```
다만 근본적으로 추가적으로 클래스 추가 시 정의에 추가해야한다는 문제점을 해결하기 위해서는 class prop의 type을 string으로 변경해야 합니다. 하지만 이 경우 오히려 휴먼 에러가 발생할 가능성만 올려줄 것 같아 배제하였습니다. 

이 외에 scss에서 선언된 class들을 가져다가 type으로 사용할 수 있을 지 찾아봤는데 이건 불가능한 것 같더라구요😅 그렇기에 이 방안이 최선이라고 생각합니다.

BangDori replied:
음 그럼 우선 버튼 스타일이 추가될 때마다, 추가해주는 걸로 하고 추후에 너무 많은 타입이 생기게 되었을 때 이에 대해 이야기 나눠보면 좋을 것 같습니다.

BangDori reviewed (changes requested):
FSD Architecture에서는 Public API를 이용하여 필요한 부분만을 외부로 추출하고, 내부 정보는 숨기는 걸로 알고있습니다.

근데 현재, shared/ui 부분에 Public API 적용 부분이 빠져있는 것 같아서 Request Changes 하였습니다.

- 참고: https://feature-sliced.design/docs/reference/public-api

Legitgoons committed:
feat: ProfileImage 없을 경우 기본 이미지 출력하도록 구현

Legitgoons committed:
feat: 사용자의 Union Type 인지를 돕기 위해 ActiveStyle, BasicStyle Union Type 분리

Legitgoons committed:
feat: BasicButton 컴포넌트 default styleClass 추가

따로 style class가 없는 button을 고려했기에 class 내용을 작성하지 않았습니다.
styleClass prop를 optinal하게 두는 방안도 고려했습니다.
하지만 어떤 styleClass를 사용하는지 명시적으로 작성하는 것이 실수를 줄이기에 더 유리하다고 생각합니다.

Legitgoons committed:
Merge branch 'main' into feature/PW-251-shared-components

Legitgoons committed:
feat: public api를 통해 export하기 위해 화살표 함수 컴포넌트로 수정

Legitgoons committed:
feat: public api를 사용해 export 하도록 각 ui폴더에 index.ts 구현

Legitgoons committed:
feat: reset.scss 코드 추가 및 이름 수정

명확한 역할을 위해 global.scss -> reset.scss로 수정

Legitgoons committed:
feat: reset.scss를 사용하기에 normalize.css 삭제

Legitgoons committed:
feat: react portal 사용을 위해 index.html에 overlays div태그 생성

Legitgoons committed:
feat: Modal에 상위요소가 영향을 미치지 않도록 ModalOverlay 컴포넌트 구현

Modal이 상위요소의 위치에 따라 영향을 받지 않도록 createPortal을 사용한 ModalOverlay 컴포넌트를 구현했습니다.

Legitgoons committed:
feat: Modal들이 ModalOverlay를 사용하도록 수정

Legitgoons committed:
feat: BottomSheetModal을 shared/ui/modal로 이동, ModalOverlay를 사용하도록 수정

Legitgoons committed:
fix: RootLayout에서 FNB를 public api로 사용하도록 수정

Legitgoons committed:
fix: followButton mixin 잘못 명시한 class 수정

Legitgoons committed:
fix: bsm-option에 누락된 border-radius 속성 추가

Legitgoons committed:
Merge branch 'feature/PW-251-shared-components' of https://github.com/CollaBu/pennyway-client-webview into feature/PW-251-shared-components

BangDori commented on src/shared/ui/profile/Profile.tsx:
```diff
@@ -0,0 +1,21 @@
+}
+
+export const Profile = ({ profileImage, name, content }: ProfileProps) => {
+  const profileImg = profileImage === '' ? 'defalut-img' : profileImage;
```
P3: `profileImg`는 로직이라기 보다는 그 자체로 View 역할을 하고 있다고 생각을 합니다. 그래서 View를 담당하는 `return`에 추가시켜도 될 것 같아요. 그리고 현재는 `default-img` 변수로 되어있는데, 추후 이미지로 변경될 예정인건가요?

Legitgoons replied:
넵 그렇게 수정하겠습니다. `default-img` 변수는 당장 기본 이미지가 정해지지 않았기에 우선 간단하게 로직만 작성해두었습니다.

BangDori commented on src/shared/ui/modal/ModalOverlay.tsx:
```diff
@@ -0,0 +1,32 @@
+    <>
+      {createPortal(<Backdrop onClick={onClose} />, portalElement)}
+      {createPortal(
+        <figure className={styleClass}>{children}</figure>,
```
P4: 모달창을 `figure` 태그로 나타내신 이유가 있으실까요?

Legitgoons replied:
modal의 의미를 잘 나타낼 태그를 찾다가 적용해봤습니다. 하지만 다시 찾아보니 modal을 그렇게 잘 나타낼 수 있다는 생각은 들지 않네요. 

다시 찾아보니 dialog 태그가 가장 적절하지만, 이걸 사용하니 modal창이 아예 안보입니다 ㅎㅎ;; 고로 section 태그를 사용하도록 하겠습니다.

BangDori commented on src/shared/ui/button/index.ts:
```diff
@@ -0,0 +1,4 @@
+export { ActiveButton } from './ActiveButton';
+export type { ActiveStyle } from './ActiveButton';
+export { BasicButton } from './BasicButton';
+export type { BasicStyle } from './BasicButton';
```
P4: 타입을 따로 `export type`으로 설정해주신 이유가 있나요? 그냥 `export`로 한 번에 묶어도 문제없지 않나요?

Legitgoons replied:
저도 처음엔 한번에 묶어서 내보내려고 했는데, 이런 에러가 발생하더라구요. `Re-exporting a type when 'isolatedModules' is enabled requires using 'export type'.` 

찾아보니 프로젝트 내 각각의 소스코드 파일을 모듈로 만들기를 강제하는 `isolatedModules`옵션이 켜져있어서 였습니다. 옵션을 끄기보다는 `export type`로 명시해주는게 나을 듯 해서 이렇게 작성했습니다.

Legitgoons replied:
그런데 지금 문득 생각 든 게, 다른 파일에서 style type을 사용하지 않을 것 같은데 애초에 export를 안하는게 맞을 것 같네요 ㅎㅎ;; 수정하겠습니다.

BangDori replied:
넵!

BangDori commented on src/shared/styles/reset.scss:
P5: 오호, 이렇게 설정해두기만 하면 페이지 로드될 때 마다 자동으로 `reset` 되는건가요??

Legitgoons replied:
넵 기존의 방식은 제 의도와는 다르게 모든 style이 reset되지 않아서 이 방법으로 수정했습니다.

BangDori replied:
기존의 방식이라 함은 혹시 어떤 방식을 말하시는 건지 여쭤봐도 될까요?

Legitgoons replied:
Normalize를 사용했었습니당!
https://velog.io/@jewon119/01.CSS-기초-Reset-CSS

BangDori reviewed (commented):
아래 이미지처럼 ui 폴더 자체에 대한 public api도 적용되어야 할 것 같아요!

![image](https://github.com/CollaBu/pennyway-client-webview/assets/44726494/cd78799e-b4b1-457a-b2b7-e035bad02d6c)


참고: https://github.com/noveogroup-amorgunov/nukeapp/blob/main/src/shared/ui/index.ts

Legitgoons committed:
feat: modal의 의미를 더 잘 나타낼 수 있는 section태그로 수정

Legitgoons committed:
feat: ProfileImage 컴포넌트 분리

Legitgoons committed:
feat: 잘못 export한 button type들 삭제

Legitgoons committed:
feat: shared/ui 전체의 index.ts 파일 생성

BangDori commented on src/shared/ui/profile/ProfileImage.tsx:
P3: 제가 말한 의도가 제대로 전달되지 못한 거 같아요! 제가 말씀 드린 건 `profileImg`가 결국 View를 위한 변수이니 삼항 연산자 혹은 논리 연산자를 이용해서 `img` 태그에 바로 적용하자는 의미였어요.

```tsx
export const Profile = ({ profileImage, name, content }: ProfileProps) => {
  return (
    <div className='profile'>
      <img src={profileImage || 'default-image'} className='profile-image' />;
      <div className='name-section'>
        <h5 className='name b1semi'>{name}</h5>
        <p className='content b3md'>{content}</p>
      </div>
    </div>
  );
};
```

아이고 혼란을 드린 것 같네요. 죄송합니다

Legitgoons replied:
아 ㅋㅋㅋㅋ 제가 오해했네요 ㅎㅎ; 다시 수정했습니다!

BangDori reviewed (commented):
`Profile` 컴포넌트에서 제가 말한 의도가 제대로 전달되지 못한 거 같아요!

죄송해요. 마지막으로 확인 부탁드릴게요.

Legitgoons committed:
feat: ProfileImage / Profile 컴포넌트 병합

Legitgoons committed:
fix: shared/ui/index 수정

BangDori reviewed (approved):
공통 컴포넌트에 대한 내용이 생각보다 너무 많았네요..

의찬님 코드 리뷰하면서 PR을 되도록 이면 간략하게, 하나의 기능에 집중해서 하는 것이 얼마나 중요한지에 대해 깨달을 수 있었던 시간이였던 것 같습니다 ㅎㅎ 

저어어엉ㅇ~~~말 고생많으셨습니다!! 🔥🔥🔥🔥🔥
